### ISSUE TITLE
Release: Merge Upstream Changes

### ISSUE THREAD

**Issue by flatcap** [neomutt/neomutt#1842]
Currently, there are 18 unmerged commits.

- [ ] https://github.com/neomutt/upstream-mutt/commit/1600dd18 Rename browser fields to display_name and full_path
- [ ] https://github.com/neomutt/upstream-mutt/commit/0fa71030 Add $browser_sticky_cursor default set
- [ ] https://github.com/neomutt/upstream-mutt/commit/2e6d4903 Convert remaining mutt_encode_path() call to use BUFFER
- [ ] https://github.com/neomutt/upstream-mutt/commit/773166e9 Change browser <display-filename> to show full_path
- [ ] https://github.com/neomutt/upstream-mutt/commit/a4b53e19 Improve sidebar indentation and shortpath behavior
- [ ] https://github.com/neomutt/upstream-mutt/commit/a6bccbe1 Memcpy header cache fetch values to ensure alignment
- [ ] https://github.com/neomutt/upstream-mutt/commit/fe69e4d1 Add sticky browser behavior for sorting
- [ ] https://github.com/neomutt/upstream-mutt/commit/0e32b977 Add new browse-mailboxes function in index and pager
- [ ] https://github.com/neomutt/upstream-mutt/commit/3f911681 Create multipart/mixed helpers
- [ ] https://github.com/neomutt/upstream-mutt/commit/0e566a03 Add ability to generate multipart/alternative using a filter script
- [ ] https://github.com/neomutt/upstream-mutt/commit/29167ecf Block multipart type alternatives generation
- [ ] https://github.com/neomutt/upstream-mutt/commit/c64baf93 Change $write_bcc to default off
- [ ] https://github.com/neomutt/upstream-mutt/commit/183c4687 Add markdown2html contrib script
- [ ] https://github.com/neomutt/upstream-mutt/commit/5835f4c0 Remove nonreentrant calls from exit_handler()
- [ ] https://github.com/neomutt/upstream-mutt/commit/aae2d87a Add $count_alternatives to count attachments inside alternatives
- [ ] https://github.com/neomutt/upstream-mutt/commit/e850e89f allow autoview preview of multipart/alternative content
- [ ] https://github.com/neomutt/upstream-mutt/commit/06698649 Change $send_multipart_alternative to run in batch mode on ask-yes
- [ ] https://github.com/neomutt/upstream-mutt/commit/24ca73ae Fix $fcc_attach to not prompt in batch mode

Exporting them to patch-form first makes it easier to merge them.
As patches, you can perform search/replace to NeoMutt-ify them.

- `git format-patch HASH^..HASH`
- Tidy commit subject
  - Remove references to their tracker
  - Remove trailing `.`
  - Shorten to 50 chars, if overly long
- Tidy the message body
- Tidy the patch
  - Search/replace Mutt structs with NeoMutt structs, etc
  - Add a credit for yourself -- this is time-consuming work
    `Co-authored-by: Richard Russon <rich@flatcap.org>`
- Apply patch: `git am PATCH`

After each patch applied, run `clang-format` on the code to tidy it up.
If the patch changes the docs, run `make validate-doc` to check them.

## Merge

Once all the changes are committed, check your work and test that NeoMutt still runs.
Finally, merge your changes:

```
git checkout master
git commit --no-ff upstream --message "upstream fixes" --message "Upstream commit: HASH"
```

The second message mentions the latest upstream commit hash that you've merged.
This helps the next person merging upstream commits.

**Comment by Austin-Ray** [comment]
I'll start merging the upstream changes into our repo.

**Comment by flatcap** [comment]
I've sorted out all my 'transform' scripts and dumped them in a gist - https://gist.github.com/flatcap/3472d03959910e1db95a6633c25b4671
They date from different eras.  Some may be more useful than others.
Also, some vim scripts use [vim-abolish](https://github.com/tpope/vim-abolish) `%S` replacements.
 
[rename-structs.sed](https://gist.github.com/flatcap/3472d03959910e1db95a6633c25b4671#file-rename-structs-sed) should be helpful.

**Comment by flatcap** [comment]
(slowly continuing my workflow...)
After `git format-patch`, scripts and hacking, I `git am 0001.patch` (which will fail).
Next, I `patch -p1 < .git/rebase-apply/patch`
Some hunks will succeed, many will fail, but that'll create `.rej` files to work through.

**Comment by Austin-Ray** [comment]
I'm tracking all my work in `devel/merge-upstream`.

My plan of attack is:
  1. Apply all relevant patches
  2. Ensure neomutt builds/functions
  3. Edit commit titles to classify changes (e.g. prefix `fix:`)
  4. Open PR with all upstream changes

How does this sound?

**Comment by flatcap** [comment]
I've rebased your branch, fixed the build:

- 3ae9cf07c Properly revert on an autocrypt postpone failure
- feb280f23 Fix memory leak when attaching messages
- aa92e5c97 fix hdr

and worked through one more of the commits: "Re-enable and cleanup format-flowed space stuffing"

- 809b7f967 PATCH
- 5dfc7b789 tidy patch
- 59d146ab2 Re-enable and cleanup format-flowed space stuffing
- a44340301 fix build
- 9278ed942 clang-format
- 905033088 tidy code

Here are all the steps I took:
- `git format-patch` to generate the upstream patch
- apply all the 'z_*.sed' scripts in the gist
- `git am 0001*` -- try to apply the patch (all chunks failed)
  This creates the template for the commit (all the metadata)
- `patch -p1 < .git/rebase-apply/patch` -- apply manually
  This creates a handful of `.rej` (reject) files
- Apply the reject files manually in Vim
  Most are fairly simple, `rfc3676.c` needed a bit of care
  I frequently `diff` the current file, `master` and the upstream version (for guidance)
- `git commit` the changes.  this'll use the patch's metadata
- Compile -- failed miserably
- Fix the build - just the minimum to get it to compile, for now
  (Lots of library functions and constants have been renamed)
- `clang-format` - tidy the files that've changed
- tidy the code to NeoMutt standards: variable naming, `()`s, Doxygen comments
- lots of test builds
- doxygen test build

When I'm finally satisified, I'll squash all the commits (and delete the PATCH commits).

**Comment by flatcap** [comment]
Rebase; squashed.  Some more Buffer changes performed; summary updated.

**Comment by flatcap** [comment]
**Merged**:
- ~https://github.com/neomutt/upstream-mutt/commit/4c2f7c70 Fix inotify configure test~                                                                              
- https://github.com/neomutt/upstream-mutt/commit/af47e694 Properly revert on an autocrypt postpone failure    
- https://github.com/neomutt/upstream-mutt/commit/c28cba77 Fix memory leak when attaching messages    
- https://github.com/neomutt/upstream-mutt/commit/ad72ec6c Convert bcache to use buffer pools    
- https://github.com/neomutt/upstream-mutt/commit/192a6d0e Convert mutt_complete() to use the buffer pool    
- https://github.com/neomutt/upstream-mutt/commit/4f89f530 Convert compose check_attachments() to use buffer pool    
- https://github.com/neomutt/upstream-mutt/commit/4050cc43 Fix mutt_save_message() "enter_fname()" to use buffer version    
- https://github.com/neomutt/upstream-mutt/commit/1ae6e10f Remove nonull check from _mutt_enter_fname()    
- https://github.com/neomutt/upstream-mutt/commit/ecd0d6ed Create mutt_buffer_get_field()    
- https://github.com/neomutt/upstream-mutt/commit/5da50b4d Convert buffer callers of mutt_get_field() to use buffer version    
- https://github.com/neomutt/upstream-mutt/commit/b4cd9852 Convert compose menu fname variable to buffer    
- https://github.com/neomutt/upstream-mutt/commit/0637998b Convert compress setup_paths() to use buffer    
- https://github.com/neomutt/upstream-mutt/commit/4d2f645a Fix memory leak in compose write-fcc function

**Comment by flatcap** [comment]
**Merged**:

- https://github.com/neomutt/upstream-mutt/commit/4959e8d Re-enable and cleanup format-flowed space stuffing
- https://github.com/neomutt/upstream-mutt/commit/7949149 Remove manual note about space-stuffing only once
- https://github.com/neomutt/upstream-mutt/commit/aafe98b Reuse the static space_stuff function for unstuffing
- https://github.com/neomutt/upstream-mutt/commit/fc0c2be Convert crypt-gpgme to use BUFFER for tempfiles
- https://github.com/neomutt/upstream-mutt/commit/13965f7 Convert crypt_extract_keys and signed handler to use buffers
- ~https://github.com/neomutt/upstream-mutt/commit/44349c1 Convert mutt_error_history to use buffer pool~
- https://github.com/neomutt/upstream-mutt/commit/a4f9923 Convert edit_one_message() to use buffer pool
- https://github.com/neomutt/upstream-mutt/commit/34ad213 Convert run_decode_and_handler() to use buffer pool
- https://github.com/neomutt/upstream-mutt/commit/84c1b64 Convert hcache db4 lockfile to buffer
- https://github.com/neomutt/upstream-mutt/commit/a736d4a Convert hcache_open to use buffer pool
- https://github.com/neomutt/upstream-mutt/commit/48cf109 Fix built-in pager checks for help and attachments
- https://github.com/neomutt/upstream-mutt/commit/dafb1f3 Convert mutt_edit_headers() to use buffer pool
- https://github.com/neomutt/upstream-mutt/commit/bb62f17 Convert mutt_help() to use buffer for tempfile
- ~https://github.com/neomutt/upstream-mutt/commit/43bc959 Convert shrink_histfile() tempfile to use buffer pool~
- https://github.com/neomutt/upstream-mutt/commit/3f5b022 Convert save-hook and fcc-hook to use buffer pool internally
- https://github.com/neomutt/upstream-mutt/commit/c65c12b Fix mutt_write_mime_body() application/pgp-encrypted handling
- https://github.com/neomutt/upstream-mutt/commit/2e0fc33 Convert imap_open_new_message() to use buffer pool
- https://github.com/neomutt/upstream-mutt/commit/dcdbdfe Convert imap fetch_new and fetch_message tempfiles to buffer pool
- https://github.com/neomutt/upstream-mutt/commit/f055ff5 Convert imap_cachepath() and callers to use buffers
- ~https://github.com/neomutt/upstream-mutt/commit/d430f37 Change message cache id to use SHORT_STRING~
- https://github.com/neomutt/upstream-mutt/commit/49c9780 Convert parse_path_(un)list to use buffer pool for path
- https://github.com/neomutt/upstream-mutt/commit/46223bc Convert start_debug() to use buffer pool

**Comment by flatcap** [comment]
@Austin-Ray 
Following your suggestion on IRC, which do you prefer:

- `Upstream-commit: 1600dd18`
  Just a quick reference

- `Upstream-commit: neomutt/upstream-mutt@1600dd18`
  Short link (GitHub will auto-link this)

- `Upstream-commit: https://github.com/neomutt/upstream-mutt/commit/1600dd189b064341b6579d8b07d0bf5f6ad80f8d`
  Explicit URL (or we could link to Mutt's GitLab)

Newer versions of `git-log` can print/filter by 'trailers' of dashed-words-colon-space.

**Comment by Austin-Ray** [comment]
@flatcap : I prefer the third option. A fully-qualified URL makes it easier to access the commit. Linking to Mutt's GitLab seems like a good idea.

**Comment by flatcap** [comment]
**Merged**:

- neomutt/upstream-mutt@472324f8 add fallback for inotify_init1
- neomutt/upstream-mutt@36ee8d39 Convert hcache fetch and store raw to use buffer pool
- neomutt/upstream-mutt@fb1686b6 Convert hcache delete and open calls to use buffer pool
- neomutt/upstream-mutt@8351c060 Convert mutt_parse_hook() to use buffer pool
- neomutt/upstream-mutt@fb2487c2 Convert set/reset and source_rc to use buffers for paths
- neomutt/upstream-mutt@e4c176b3 Convert main() to use buffers for paths
- neomutt/upstream-mutt@937a1eb1 Clear connection input buffer when closing
- neomutt/upstream-mutt@5572a302 Convert mbox_sync_mailbox() to use buffer pool
- neomutt/upstream-mutt@140cd885 Change ssl_init to use buffer pool for rand filename
- ~neomutt/upstream-mutt@48847904 Removed unused mutt_expand_link()~
- neomutt/upstream-mutt@0944e898 Convert mutt_save_confirm() to use buffer pool
- neomutt/upstream-mutt@23bfc45f Convert mx_close_mailbox() to use buffer pool
- ~neomutt/upstream-mutt@f566f734 Convert msg_search to use buffer~
- neomutt/upstream-mutt@30b1ff48 Convert pgp_app_handler to use buffer pool
- neomutt/upstream-mutt@c1f9f00f Convert pgp_check_traditional_one_body() to use buffer pool
- neomutt/upstream-mutt@e2722551 Convert pgp_verify_one() to use buffer pool
- neomutt/upstream-mutt@f3bfb7d1 Convert pgp_extract_keys_from_attachment() to use buffer pool
- neomutt/upstream-mutt@e383f5c9 Convert pgp_decrypt_part() to use buffer pool
- ~neomutt/upstream-mutt@aa66b414 Convert pgp_decrypt_mime() to use buffer pool~
- ~neomutt/upstream-mutt@48a34377 Convert pgp_encrypted_handler() to use buffer pool~
- neomutt/upstream-mutt@38291748 Convert pgp_sign_message to use buffer pool
- neomutt/upstream-mutt@978cdaf7 Convert pgp_encrypt_message() to use buffer pool
- neomutt/upstream-mutt@782bcd7d Convert pgp_traditional_encryptsign() to use buffer pool
- neomutt/upstream-mutt@e3408a00 Convert pgp_select_key() to use buffer pool
- ~neomutt/upstream-mutt@15f89f8e Remove make_key_attachment parameter~
- neomutt/upstream-mutt@26e89164 Convert pgp_make_key_attachment() to use buffer pool
- ~neomutt/upstream-mutt@5b53290a Convert pgp_find_hash to use buffer pool~
- ~neomutt/upstream-mutt@08480fa8 Convert pop_read_header() to use buffer pool~
- neomutt/upstream-mutt@d8b6f94e Convert pop_fetch_message() to use buffer pool
- neomutt/upstream-mutt@80bc61db Convert mutt_num_postponed() maildir check to use buffer pool
- neomutt/upstream-mutt@1f71fad9 Convert attach_fmt %f to use buffer pool
- neomutt/upstream-mutt@0b59dc75 Convert attach_forward_bodies to use buffer pool
- neomutt/upstream-mutt@8d12d900 Convert attach_forward_msgs() to use buffer pool
- neomutt/upstream-mutt@6380cbd6 Convert mutt_attach_reply() to use buffer pool
- neomutt/upstream-mutt@8997c55f Convert rfc1524_mailcap_lookup() to use buffer pool for path
- neomutt/upstream-mutt@2fcde005 Convert send_message() to use buffer pool
- neomutt/upstream-mutt@c5d770ea Convert fcc to a buffer

**Comment by flatcap** [comment]
**Merged**:

- neomutt/upstream-mutt@71d6dbea Update mime fields when piping a message with $pipe_decode set
- ~neomutt/upstream-mutt@9506db1e Reduce line buffer size in mx_get_magic()~
- neomutt/upstream-mutt@a327386c Stable branch quick fix for pager change-mailbox push/exec bug
- neomutt/upstream-mutt@b15e8659 Remove menu->menu hack when redirecting pager ops through index
- neomutt/upstream-mutt@e750e39c Clean up pager change folder aborts to return to pager
- neomutt/upstream-mutt@12cd2d52 Add typelen parameter to rfc1524_mailcap_lookup()

**Merged**:

- neomutt/upstream-mutt@671e653c Convert mutt_message_to_7bit() to use buffer
- neomutt/upstream-mutt@3565f686 Convert transform_to_7bit() to use buffer pool
- neomutt/upstream-mutt@8de05d5f Convert send_msg() $sendmail_wait to use buffer pool
- neomutt/upstream-mutt@58a81897 Convert _mutt_bounce_message to use buffer pool
- neomutt/upstream-mutt@7e29eb23 Convert mutt_write_fcc() to use buffer pool
- neomutt/upstream-mutt@7c81cd61 Convert smime_get_candidates() to use buffer pool
- neomutt/upstream-mutt@d4c00c67 Convert smime_verify_sender() and helpers to use buffer pool
- neomutt/upstream-mutt@07e32b7a Convert smime_invoke_import() and helpers to use buffer pool
- neomutt/upstream-mutt@ce55dbbf Convert smime_build_smime_entity() to use buffer pool
- neomutt/upstream-mutt@56856d8e Convert smime_sign_message() to use buffer pool
- neomutt/upstream-mutt@05805ddd Convert smime_verify_one() to use buffer pool
- neomutt/upstream-mutt@ccf3f493 Convert smime_handle_entity() to use buffers
- neomutt/upstream-mutt@e9ab7539 Convert smime_decrypt_mime() to use buffer pool

**Comment by flatcap** [comment]
**Merged**:

- https://github.com/neomutt/upstream-mutt/commit/4cec4b63 mutt_parse_content_type: work around annoying charset behaviour from Outlook
- https://github.com/neomutt/upstream-mutt/commit/ab9b7b08 Ensure APP_PGP flag is enabled for autocrypt
- https://github.com/neomutt/upstream-mutt/commit/ef7e5e42 Fix sigwinch race in the index and menu processor
- https://github.com/neomutt/upstream-mutt/commit/8ae8394f Disable state messages for attachments when forwarding
- https://github.com/neomutt/upstream-mutt/commit/c64baf93 Change $write_bcc to default off
- https://github.com/neomutt/upstream-mutt/commit/afb14d62 Update copyrights
- https://github.com/neomutt/upstream-mutt/commit/09dfed3f Fix unused result warning in enter.c
- https://github.com/neomutt/upstream-mutt/commit/f098ac2d Fix unused result issue in help.c
- https://github.com/neomutt/upstream-mutt/commit/20c7b0d8 Fix unused result issue in query.c

**Comment by madduck** [comment]
Wondering what the status is here. I would love to finally migrate to neomutt altogether, but I definitely want the multipart/alternative generation patch. https://github.com/neomutt/upstream-mutt/commit/0e566a03

If there's anything I can help with that's blocking this, please let me know.

**Comment by dcpurton** [comment]
@madduck The multipart handling in NeoMutt has recently been improved. Although it works in a different way to Mutt, it may suit you.

Rather than configuration variables, you can set up a macro that pipes your text/plain part to a filter, attaches the result and constructs a multipart group.

See https://github.com/neomutt/neomutt/discussions/3236 for the basic idea.

I use this to send multipart/alternative emails with good control and success.

**Comment by roccoblues** [comment]
This is most certainly outdated. Closing.
